src/request.rs

```diff
@@ -238,27 +267,176 @@ impl RequestBuilder {
     }
 }
 
+impl MultipartRequestBuilder {
+    /// Constructs a new request.
+    pub fn new(request_builder: RequestBuilder) -> MultipartRequestBuilder {
+        MultipartRequestBuilder { request_builder: request_builder, params: None, files: None }
+    }
+
+    /// Add a `Header` to this Request.
+    ///
+    /// ```rust
+    /// # use reqwest::Error;
+    /// #
+    /// # fn run() -> Result<(), Error> {
+    /// use reqwest::header::UserAgent;
+    /// let client = reqwest::Client::new()?;
+    ///
+    /// let res = client.get("https://www.rust-lang.org")?
+    ///     .header(UserAgent("foo".to_string()))
+    ///     .send()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn header<H>(&mut self, header: H) -> &mut MultipartRequestBuilder
+        where H: ::header::Header + ::header::HeaderFormat
+    {
+        self.request_builder.request_mut().headers.set(header);
+        self
+    }
+
+    /// Add a set of Headers to the existing ones on this Request.
+    ///
+    /// The headers will be merged in to any already set.
+    pub fn headers(&mut self, headers: ::header::Headers) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().headers.extend(headers.iter());
+        self
+    }
+
+    /// Enable HTTP basic authentication.
+    pub fn basic_auth<U, P>(&mut self, username: U, password: Option<P>) -> &mut MultipartRequestBuilder
+        where U: Into<String>,
+              P: Into<String>
+    {
+        self.request_builder.header(::header::Authorization(::header::Basic {
+                                                username: username.into(),
+                                                password: password.map(|p| p.into()),
+                                            }));
+        self
+    }
+
+    /// Set the request body.
+    pub fn body<T: Into<Body>>(&mut self, body: T) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().body = Some(body.into());
+        self
+    }
+
+    /// Add parameters to the multipart/form-data `Request`.
+    pub fn params(&mut self, params: Params) -> &mut MultipartRequestBuilder {
```

Comment:
It might be nice if the `params` method worked similar to the `form` method on the regular build. That is, that instead of `Params`, this were generic over `Serialize`.

```rust
pub fn params<T: Serialize>(&mut self, params: &T) -> ::Result<&mut MultipartRequestBuilder> {
}
```

To make that easier, it'd require some `MultipartSerializer::new(boundary).serialize(params)`...

-----

Alternatively, we could look at the `FormData` API in web browsers. You call `append` a bunch of times, adding form fields. Is it better for users call `req.param("username", "sean").param("password", "pass")`, or to be generic over `Serialize`? I lean towards being generic, as it is the most flexible...